### PR TITLE
Implement the new consent logic to account for side-loaded studies

### LINKED ISSUE #352: Deal with consent and side-loaded studies
The proposal lives here: https://docs.google.com/document/d/1E2HUP07WCdlj9I-tH9PoCUU3cOrc0JWjEw2oxnpszmE/edit#

### PR DESCRIPTION
**Pull request by Dexterp37**
Fixes #352 #396

This slightly changes the study enrolment logic to implement a staged approach, to prevent side-loaded studies to suddenly start running without user approval. How does this work?

1. Whenever a study is consented and installation is triggered from the core add-on UI, we record a "pending consent" (i.e. user consented but the study is not yet installed).
2. When a study is installed we detect it and check if we had a prior "pending consent" for it. If so, join the study.
3. Joined installed studies are marked as such in the UI. If a study is installed, but not joined, they get uninstalled at the earliest
   opportunity by the core add-on.

Note that pending consent is cleared whenever the core add-on is initialized again, so that pending consent is not retained across browser restarts.

**Important**: I'm planning on adding integration tests either later in this PR (if the approach looks good to you) or in a separate PR.

Flagging @rhelmer for the full review, @hamilton specifically for the Svelte changes (especially the ones in the storybook).

Checklist for reviewer:

- [x] The description should reference a bug or github issue, if relevant.
- [x] There must be a [`CHANGELOG.md`](./CHANGELOG.md) entry for any non-test change.
- [ ] Any change to the NPM commands must be carefully reviewed to make sure it won't break the Add-ons pipeline.
- [ ] Any version increase must follow the [release process](./docs/RELEASE_PROCESS.md).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 [Full changelog](https://github.com/mozilla-rally/core-addon/compare/v0.9.0...master)
 
+* [#392](https://github.com/mozilla-rally/rally-core-addon/pull/392): Implement the new consent logic to prevent running side-loaded studies.
```
Comment (rhelmer) [inline CHANGELOG.md:5]: Nice job catching this issue btw :)

File: core-addon/Core.js [hunk 2]
```diff
@@ -129,11 +129,30 @@ module.exports = class Core {
...
+      //    consented but the study is not yet installed).
+      // 2. When a study is installed we hit _this code_ and check if we
+      //    had a prior "pending consent" for it. If so, join the study.
+      // 3. Joined installed studies are marked as such in the UI. If a study
+      //    is installed, but not joined, they get uninstalled at the earliest
+      //    opportunity by the core add-on.
+      //
+      // Note that pending consent is cleared whenever the core add-on is initialized
+      // again, so that pending consent is not retained across browser restarts.
+      let wasConsented = await this._storage.removePendingConsent(info.id);
```
Comment (rhelmer) [inline core-addon/Core.js:150]: I think `hasConsent` feels a little more natural but nbd.

**Review by rhelmer — COMMENTED** [review]
This looks good to me overall, I'd like to test it a bit and take a look one you add integration tests (I agree that they probably make sense here).

File: CHANGELOG.md (hunk 1 above)
Comment (Dexterp37) [reply CHANGELOG.md:5]: To be honest, I didn't! @knowtheory did :-)

**Comment by Dexterp37** [comment]
> 
> 
> This looks good to me overall, I'd like to test it a bit and take a look one you add integration tests (I agree that they probably make sense here).

Thanks for the first pass @rhelmer 

In order to add test coverage I needed to work on "generating studies" on the fly. Sorry, this bloated the PR a bit :(

**Comment by rhelmer** [comment]
> > This looks good to me overall, I'd like to test it a bit and take a look one you add integration tests (I agree that they probably make sense here).
> 
> Thanks for the first pass @rhelmer
> 
> In order to add test coverage I needed to work on "generating studies" on the fly. Sorry, this bloated the PR a bit :(

No worries! This is great to have, it's the method webextension and add-ons tests in Firefox tend to use and it works well there.

**Review by hamilton — COMMENTED** [review]
This looks good from Svelte perspective.

**Review by rhelmer — APPROVED** [review]
These test improvements are excellent!

**Pull request merged**